### PR TITLE
feat(router): add incoming header request logs

### LINKED ISSUE #4182: [FEATURE] VIPPS API (Norwegian)
### Feature Description

Use vipps as payment.
https://developer.vippsmobilepay.com/docs/recommended-flows/online/

### Possible Implementation

User can choose Vipps as payment and will be able to pay with this

### Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

None

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

added incoming header request logs

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

[closes #4182 tps://github.com/juspay/hyperswitch-cloud/issues/4182](https://github.com/juspay/hyperswitch-cloud/issues/4182)

## How did you test it?

Try Payments create and look for BEGIN REQUEST tag in logs, that log should have headers logged along with whitelisted keys as unmasked.

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_34WfMZpEPjMWHfBnwVMZIIu0hy7GhCR7vkfvZflWZ3tjww8pR6Yoloj4VeSQHwZZ' \
--data-raw '{
    "amount": 250,
    "currency": "USD",
    "confirm": true,
    "amount_to_capture": 250,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "routing": {
        "type": "single",
        "data": "nmi"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4000000000002503",
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "999"
        }
    },
    "billing": {
        "address": { 
            "zip": "94122",
            "first_name": "sahkal",
            "last_name": "poddar"
        }
    }
}

'
```

Screenshot 2024-03-04 at 5 46 19 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/configs/settings.rs [hunk 1]
```diff
@@ -115,6 +115,13 @@ pub struct Settings<S: SecretState> {
     pub events: EventsConfig,
     #[cfg(feature = "olap")]
     pub connector_onboarding: SecretStateContainer<ConnectorOnboarding, S>,
+    pub white_list_incoming_header: WhiteListIncomingHeaders,
```
Comment (SanchithHegde) [inline crates/router/src/configs/settings.rs:118]: Can we call this "allow list"?

File: config/development.toml
Comment (SanchithHegde) [inline config/development.toml:1]: Can you make this change in `config/docker_compose.toml` and `loadtest/config/developlment.toml`?

File: crates/router/src/services/api.rs [hunk 2]
```diff
@@ -1824,13 +1843,12 @@ pub fn build_redirection_form(
                     const threeDSsecureInterface = threeDS.createUI(options);
 
                     threeDSsecureInterface.on('challenge', function(e) {{
-                        document.getElementById('loader-wrapper').style.display = 'none';
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:1827]: Can we revert this change?

File: crates/router/src/services/api.rs [hunk 3]
```diff
@@ -1824,13 +1843,12 @@ pub fn build_redirection_form(
...
-                    }});
-
-                    threeDSsecureInterface.on('complete', function(e) {{
                         var responseForm = document.createElement('form');
                         responseForm.action=window.location.pathname.replace(/payments\\/redirect\\/(\\w+)\\/(\\w+)\\/\\w+/, \"payments/$1/$2/redirect/complete/nmi\");
                         responseForm.method='POST';
+                    }});
+
+                    threeDSsecureInterface.on('complete', function(e) {{
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:1851]: Same, let's revert this change.

**Review by Sakilmostak — DISMISSED** [review]

File: crates/router/src/services/api.rs (hunk 2 above)
Comment (sahkal) [reply crates/router/src/services/api.rs:1827]: This change is not related to my PR somehow this change is showing up, though this portion of my file is in Sync with main. Have tested NMI where this change lies also verified and took approval from @Sakilmostak .

File: crates/router/src/services/api.rs (hunk 3 above)
Comment (sahkal) [reply crates/router/src/services/api.rs:1851]: +1

File: crates/router/src/configs/settings.rs (hunk 1 above)
Comment (SanchithHegde) [reply crates/router/src/configs/settings.rs:118]: Sounds good, even `unmasked_headers` should be fine.

Comment (sahkal) [reply crates/router/src/configs/settings.rs:118]: cool, will make the required change

File: crates/router/src/services/api.rs (hunk 2 above)
Comment (SanchithHegde) [reply crates/router/src/services/api.rs:1827]: Ideally, this shouldn't be showing up in the diff at all. Can we just pick the entire block/function from the `main` branch using the GitHub web UI and include it in your branch, if that helps?

Comment (sahkal) [reply crates/router/src/services/api.rs:1827]: i did copy paste the same UI changes from the main branch still i can see the same.

File: crates/router/src/services/api.rs [hunk 4]
```diff
@@ -1121,11 +1121,35 @@ where
 {
     let request_method = request.method().as_str();
     let url_path = request.path();
+
+    let white_listed_incoming_header_keys = state.conf().unmasked_headers.keys;
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:1125]:
```suggestion
    let unmasked_incoming_header_keys = state.conf().unmasked_headers.keys;
```

OR

```suggestion
    let unmasked_header_names = state.conf().unmasked_headers.keys;
```

**Review by SanchithHegde — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
